debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java

```diff
@@ -540,12 +552,10 @@ public JdbcConnection queryWithBlockingConsumer(String query, StatementFactory s
      */
     public JdbcConnection prepareQuery(String preparedQueryString, StatementPreparer preparer, ResultSetConsumer resultConsumer)
             throws SQLException {
-        Connection conn = connection();
-        try (PreparedStatement statement = conn.prepareStatement(preparedQueryString);) {
-            preparer.accept(statement);
-            try (ResultSet resultSet = statement.executeQuery();) {
-                if (resultConsumer != null) resultConsumer.accept(resultSet);
-            }
+        final PreparedStatement statement = createPreparedStatement(preparedQueryString);
```

Comment:
@jpechane It might be a stupid question, but is it actually safe to use `PreparedStatements``s like this? Won't we have issues if the same statement is accessed concurrently from multiple threads? Or am I missing something here?

Reply:
@gunnarmorling Technically you are right. If there were two threads using the same prepared statement then this can be the case but there is no such an occurenc ein the code.
Generally we have multiple options
* keep it as is with the given limitations
* make cache to be a threadlocal variable - would need the central bookkeeping for cleanup
* introduce real caching soultion with acquire/releease semantics
* don't do central caching at all and make the caching to be the task of the caller

I thought we can live with the limitation of the first option. If not we should follow the last one.

Reply:
Ok, taking a step back, `JdbcConnection` itself is not meant to be accessed from multiple threads at the same time, so this change here is fine. I was a bit misguided by the notion of caching, assuming it'd be done on a more global level.

> I thought we can live with the limitation of the first option.

+1

I'd suggest to do two things then:

* Add the `@NotThreadSafe` marker annotation to `JdbcConnection`
* Revert to the original usage of `computeIfAbsent()` which is fine if concurrent usage is ruled out; the usage of the locks is overkill in that light

Sorry for the forth and back, I realized too late that this all won't be shared across threads to begin with. Btw. in general let's not introduce thread-local state, I've been burnt by it fart too often, esp. given the connectors are deployed into a container-like environment (Kafka Connect), where thread-locals may be prone to leakage upon connector removal if not taking care.

Reply:
@gunnarmorling I've reverted the last commit and added the requested annotation.